document release checklist (#6986)
numba/numba: docs/source/developer/release.rst

> Numba Release Checklist Templates
> =================================

Reviewer: ```suggestion
Numba Release Process
=====================
```

The goal of the Numba release process -- from a high level perspective -- is to
publish source and binary artifacts that correspond to a given version
number. This usually involves a sequence of individual tasks that must be
performed in the correct order and with diligence. Numba and llvmlite are
commonly released in lockstep since there is usually a one-to-one mapping
between a Numba version and a corresponding llvmlite version.

This section contains various notes and templates that can be used to create a
Numba release checklist on the Numba Github issue tracker. This is an aid for
the maintainers during the release process and helps to ensure that all tasks
are completed in the correct order and that no tasks are accidentally omitted.

If new or additional items do appear during release, please do remember to add

> them to the checklist templates. Note also, that the release process itself is
> always a work in progress. This means, some of the information here may be
> outdated. If you detect this, please do remember to submit a pull-request to
> update this document.

Reviewer: ```suggestion
them to the checklist templates. Also note that the release process itself is
always a work in progress. This means that some of the information here may be
outdated. If you notice this please do remember to submit a pull-request to
update this document.
```

> All release checklists are available as Gitub issue templates, so to create a
> new release checklist, simply open a new issue and select the correct template.

Reviewer: ```suggestion
All release checklists are available as Gitub issue templates. To create a new
release checklist simply open a new issue and select the correct template.
```

Primary Release Candidate Checklist
-----------------------------------

This is for the first/primary release candidate for minor release i.e. the
first release of every series. It is special, because during this release, the
release branch will have to be created. Release candidate indexing begins at 1.

.. literalinclude:: ../../../.github/ISSUE_TEMPLATE/first_rc_checklist.md
    :language: md
    :lines: 9-

> Open a release checklist: https://github.com/numba/numba/issues/new?template=first_rc_checklist.md

Reviewer: ```suggestion
`Open a primary release checklist <https://github.com/numba/numba/issues/new?template=first_rc_checklist.md>`_.
```
Can these just be standard links?

Subsequent Release Candidates, Final Releases and Patch Releases
----------------------------------------------------------------

Releases subsequent to the first release in a series usually involves a series
of cherry-picks, the recipe is therefore slightly different.

.. literalinclude:: ../../../.github/ISSUE_TEMPLATE/sub_rc_checklist.md
    :language: md
    :lines: 9-

> Open a release checklist: https://github.com/numba/numba/issues/new?template=sub_rc_checklist.md

Reviewer: ```suggestion
`Open a subsequent release checklist <https://github.com/numba/numba/issues/new?template=sub_rc_checklist.md>`_.
```

Numba Release Checklist Templates
==================================

> The goal of the Numba release process -- from a birds eye perspective -- is to

Reviewer: ```suggestion
The goal of the Numba release process -- from a high level perspective -- is to
```

> publish source and binary artifacts that correspond to the given version

Reviewer: ```suggestion
publish source and binary artifacts that correspond to a given version
```

number. This usually involves a sequence of individual tasks that must be
performed in the correct order and with diligence. Numba and llvmlite are
commonly released in lockstep since there is usually a one-to-one mapping
between a Numba version and a corresponding llvmlite version.

This section contains various notes and templates that can be used to create a

> Numba release checklist on the Numba Github issue tracker. This is the aid for

Reviewer: ```suggestion
Numba release checklist on the Numba Github issue tracker. This is an aid for
```

the maintainers during the release process and helps to ensure that all tasks
are completed in the correct order and that no tasks are accidentally omitted.

Historically, the release checklists were copied from one release to the next.
While this worked OKish for many years, often times tasks were lost during

> repeated copy and paste. As a result this document was created to keep track of

Reviewer: Is the history needed?

Author: not sure, maybe not, it was mostly documenting my personal motivation, I can remove it if just adds overhead.

Reviewer: I'm not sure it's necessary, a good reason for this process is presented in the previous paragraph. I'd rather spend effort keeping documentation such as this accurate with respect to changes/updates, if you feel differently am happy to discuss.

Author: sure, no need for that perspective here.

all potential tasks. All templates within this document are written in Github

> Markdown and are available as issue templates on Github. This means,

Reviewer: Why are they called `.yml` as files in the doc source tree if they are markdown `.md`?

Author: Because the format of the templates is such that you write markdown for the template content and then wrap that in yaml to add metadata such as title etc...

Author: So, this piece of documentation threw me off:

```
Issue templates are stored on the repository's default branch, in a hidden .github/ISSUE_TEMPLATE directory. If you create a template in another branch, it will not be available for collaborators to use. Issue template filenames are not case sensitive, and need a .md extension. To be included in the community profile checklist, issue templates must be located in the .github/ISSUE_TEMPLATE folder and contain valid name: and about: YAML front matter.
```

From: https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates 

Will rename the files to `.md`.

Author: done

to create a new release checklist, simply open a new issue and select the
correct template.

If new or additional items do appear during release, please do remember to add
them to the checklist templates. Note also, that the release process itself is
always a work in progress. This means, some of the information here may be
outdated. If you detect this, please do remember to submit a pull-request to
update this document.

Primary Release Candidate Checklist
-----------------------------------

> This is for the first/primary release candidate for minor release. I.e. the

Reviewer: ```suggestion
This is for the first/primary release candidate for minor release i.e. the
```

first release of every series. It is special, because during this release, the

> release branch will have to be created. Release candidate indexing begins at 0.

Reviewer: ```suggestion
release branch will have to be created. Release candidate indexing begins at 1.
```
it begins at `1`? See note above about consistency.

Author: Yes, it begins at 1.

.. literalinclude:: first_rc_checklist.yml
    :language: md

>     :lines: 9-

Author: My intent here was to make the document readable and self-contained. Also the actual template contains additional "cruft" such as the yaml header which may be confusing, perhaps.

>     :lines: 9-

Author: I thought about this some more, and I would like the templates to be rendered in the documentation. It makes it easier to refer to them and to view them. I have also added a link to open such an issue checklist, for convenience.

>     :lines: 9-

Reviewer: I'm not entirely sure about this as the gains aren't yet clear to me, however I think it's worth a try to see how it goes.

> Second Releases Candidates, Final Releases and Patch Releases
> -------------------------------------------------------------

Reviewer: ```suggestion
Subsequent Release Candidates, Final Releases and Patch Releases
----------------------------------------------------------------
```

Author: Yes, one of my favorite words!

> A patch release usually involves a series of cherry-picks, so the recipe is
> slightly different.

Reviewer: ```suggestion
Releases subsequent to the first release in a series usually involves a series
of cherry-picks, the recipe is therefore slightly different.
```

A patch release usually involves a series of cherry-picks, so the recipe is
slightly different.

.. code-block::

    ## numba X.Y.Z

    * [ ] cherry-pick items from the X.Y.Z milestone into a PR
    * [ ] merge change log modifications and cherry-picks to X.Y release branch
    * [ ] https://github.com/numba/numba/pull/XXXX
    * [ ] annotated tag X.Y.Z on release branch (no `v` prefix)
    * [ ] build and upload conda packages on buildfarm (check "upload")
    * [ ] build wheels (`$PYTHON_VERSIONS`) on the buildfarm
    * [ ] upload wheels and sdist to PyPI (upload from `ci_artifacts`)
    * [ ] verify packages uploaded to Anaconda Cloud and move to `numba/label/main`
    * [ ] verify wheels for all platforms arrived on PyPi
    * [ ] verify ReadTheDocs build
    * [ ] clean up `ci_artifacts`
    * [ ] send RC/FINAL announcement email / post announcement to discourse group
    * [ ] post link to Twitter
    * [ ] post link to python-announce-list@python.org

    ### Post release

    * [ ] cherry-pick change-log modifications to main branch (`master`)
    * [ ] update release checklist template
    * [ ] ping Anaconda Distro team to trigger a build for `defaults` (FINAL ONLY)

>     * [ ] close milestone (and then close this release issue)

Reviewer: Could these just be actual templates?

Reviewer: In this directory: https://github.com/numba/numba/tree/master/.github/ISSUE_TEMPLATE

Author: Maybe. However, I am not sure we can git version control the templates then. I think keeping track of how they evolve might be useful. Maybe we can update the templates from the documentation directly?

Author: Maybe I can symlink the templates into Sphinx, let me try that.

Reviewer: The templates are versioned by virtue of them being in the repo? You can see e.g. blame from numerous changes to the bug report template here:
https://github.com/numba/numba/blame/master/.github/ISSUE_TEMPLATE/Bug_report.md

or do you mean something else?

Author: > The templates are versioned by virtue of them being in the repo? You can see e.g. blame from numerous changes to the bug report template here:
> https://github.com/numba/numba/blame/master/.github/ISSUE_TEMPLATE/Bug_report.md
> 
> or do you mean something else?

I wasn't aware that this is how tempaltes worked. I though they were simply attached to the repo in an unversioned form. Thank you for clearing this up.

Reviewer: No problem. What's also quite nice about maybe doing it this way is that you don't have to expose the template in the https://github.com/numba/numba/blob/master/.github/ISSUE_TEMPLATE/config.yml (as it's not useful to users) but I think it's still accessible directly as needed. E.g. by accessing:
```
https://github.com/numba/numba/issues/new?template=THE_RELEASE_TEMPLATE.md
```

Author: That is useful, I have included such URLs for both templates.
